c++/src/kj/compat/http.c++

```diff
@@ -1729,33 +1826,58 @@ static_assert(!fastCaseCmp<'f','O','o','B'>("FooB1"), "");
 static_assert(!fastCaseCmp<'f','O','o','B','1','a'>("FooB1"), "");
 
 kj::Own<kj::AsyncInputStream> HttpInputStreamImpl::getEntityBody(
-    RequestOrResponse type, HttpMethod method, uint statusCode,
+    RequestOrResponse type,
+    kj::OneOf<HttpMethod, HttpConnectMethod> method,
+    uint statusCode,
     const kj::HttpHeaders& headers) {
   KJ_REQUIRE(headerBuffer.size() > 0, "Cannot get entity body after header buffer release.");
 
+  auto isHeadRequest = method.tryGet<HttpMethod>().map([](auto& m) {
+    return m == HttpMethod::HEAD;
+  }).orDefault(false);
+
+  auto isConnectRequest = method.is<HttpConnectMethod>();
+
   // Rules to determine how HTTP entity-body is delimited:
   //   https://tools.ietf.org/html/rfc7230#section-3.3.3
-
   // #1
   if (type == RESPONSE) {
-    if (method == HttpMethod::HEAD) {
+    if (isHeadRequest) {
       // Body elided.
       kj::Maybe<uint64_t> length;
       KJ_IF_MAYBE(cl, headers.get(HttpHeaderId::CONTENT_LENGTH)) {
         length = strtoull(cl->cStr(), nullptr, 10);
       } else if (headers.get(HttpHeaderId::TRANSFER_ENCODING) == nullptr) {
-        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD request.
-        //   Propagate this fact with a 0 expected body length.
+        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD
+        //   request. Propagate this fact with a 0 expected body length.
         length = uint64_t(0);
       }
       return kj::heap<HttpNullEntityReader>(*this, length);
+    } else if (isConnectRequest && statusCode >= 200 && statusCode < 300) {
+      // #2 For any 2xx response to a CONNECT request... a client MUST ignore any
+      // Content-Length or Transfer-Encoding header fields received.
+      return kj::heap<HttpConnectionCloseEntityReader>(*this);
     } else if (statusCode == 204 || statusCode == 304) {
       // No body.
       return kj::heap<HttpNullEntityReader>(*this, uint64_t(0));
     }
   }
 
-  // #2 deals with the CONNECT method which is handled separately.
+  // For CONNECT requests messages, we will ignore both the Transfer-Encoding
+  // and the Content-Length.
+  //
+  // TODO(now): We need to decide if we are going to allow for CONNECT request
+  // message payloads or not. Per the spec, "A payload within a CONNECT request
+  // message has no defined semantics; sending a payload body on a CONNECT request
+  // might cause some existing implementations to reject the request."
+  //
+  // Here, we just choose to ignore any delimination of any payload and act as if
+  // whatever is received after the headers is just going to be part of the
+  // CONNECT tunnel. But is that safe? Should we instead active refuse to process
+  // CONNECT requests with a payload?
```

Comment:
Check out the comment here. We need to decide how we want to handle the possibility of a CONNECT request having a payload.

Reply:
Ok, updated to have the server check on the presence of `Content-Length` and `Transfer-Encoding` headers in the request and to reject those with a `400 Bad Request` automatically.